File: test-suite/src/test/java/io/micronaut/docs/http/client/bind/MetadataClientArgumentBinder.java
package io.micronaut.docs.http.client.bind;

//tag::clazz[]
import edu.umd.cs.findbugs.annotations.NonNull;
import io.micronaut.core.convert.ArgumentConversionContext;
import io.micronaut.core.type.Argument;
import io.micronaut.docs.http.client.bind.Metadata;
import io.micronaut.http.MutableHttpRequest;

import javax.inject.Singleton;

@Singleton
public class MetadataClientArgumentBinder implements TypedClientArgumentRequestBinder<Metadata> {

    @Override
    @NonNull
    public Argument<Metadata> argumentType() {
        return Argument.of(Metadata.class);
    }

    @Override
    public void bind(@NonNull ArgumentConversionContext<Metadata> context, @NonNull Metadata value, @NonNull MutableHttpRequest<?> request) {
        request.header("X-Metadata-Version", value.getVersion().toString());
        request.header("X-Metadata-Deployment-Id", value.getDeploymentId().toString());
    }
}
//end::clazz[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/http/client/bind/type/MetadataClientArgumentBinder.kt
package io.micronaut.docs.http.client.bind.type

//tag::clazz[]
import io.micronaut.core.convert.ArgumentConversionContext
import io.micronaut.core.type.Argument
import io.micronaut.http.MutableHttpRequest
import io.micronaut.http.client.bind.TypedClientArgumentRequestBinder
import javax.inject.Singleton

@Singleton
class MetadataClientArgumentBinder : TypedClientArgumentRequestBinder<Metadata?> {

    override fun argumentType(): Argument<Metadata?> {
        return Argument.of(Metadata::class.java)
    }

    override fun bind(context: ArgumentConversionContext<Metadata?>, value: Metadata, request: MutableHttpRequest<*>) {
        request.header("X-Metadata-Version", value.version.toString())
        request.header("X-Metadata-Deployment-Id", value.deploymentId.toString())
    }
}
//end::clazz[]
